main.go

```diff
@@ -49,206 +43,378 @@ import (
+	}
+	p2p.WaitForPeersStart()
+	p2p.WaitForSyncBlkFinish()
 
-	txPoolServer.RegisterActor(tc.NetActor, p2pPID)
-	hserver.SetNetServerPID(p2pPID)
-	hserver.SetLedgerPid(ledgerPID)
-	hserver.SetTxnPoolPid(txPoolServer.GetPID(tc.TxPoolActor))
-	hserver.SetTxPid(txPoolServer.GetPID(tc.TxActor))
-	go restful.StartServer()
+	log.Infof("P2P node init success")
+	return p2p, p2pPID, nil
+}
 
-	if consensusType != "vbft" {
-		p2p.WaitForPeersStart()
-		p2p.WaitForSyncBlkFinish()
+func initConsensus(ctx *cli.Context, p2pPid *actor.PID, txpoolSvr *proc.TXPoolServer, wallet *account.ClientImpl) (consensus.ConsensusService, error) {
+	if !config.DefConfig.Common.EnableConsensus {
+		return nil, nil
 	}
+	acc := wallet.GetDefaultAccount()
+	if acc == nil {
+		return nil, fmt.Errorf("GetDefaultAccount failed")
+	}
+	pool := txpoolSvr.GetPID(tc.TxPoolActor)
 
-	if nettypes.SERVICE_NODE_NAME != config.Parameters.NodeType {
-		log.Info("5. Start Consensus Services")
-		pool := txPoolServer.GetPID(tc.TxPoolActor)
-		consensusService, _ := consensus.NewConsensusService(acct, pool, nil, p2pPID)
-		netreqactor.SetConsensusPid(consensusService.GetPID())
-		go consensusService.Start()
-		time.Sleep(5 * time.Second)
-		hserver.SetConsensusPid(consensusService.GetPID())
-		go localrpc.StartLocalServer()
+	consensusType := strings.ToLower(config.DefConfig.Genesis.ConsensusType)
+	consensusService, err := consensus.NewConsensusService(consensusType, acc, pool, nil, p2pPid)
+	if err != nil {
+		return nil, fmt.Errorf("NewConsensusService:%s error:%s", consensusType, err)
 	}
+	netreqactor.SetConsensusPid(consensusService.GetPID())
+	hserver.SetConsensusPid(consensusService.GetPID())
+
+	go consensusService.Start()
+
+	log.Infof("Consensus init success")
+	return consensusService, nil
+}
+
+func initRpc(ctx *cli.Context) error {
+	var err error
+	exitCh := make(chan interface{}, 0)
+	go func() {
+		err = jsonrpc.StartRPCServer()
+		close(exitCh)
+	}()
 
-	log.Info("--Start the RPC interface")
-	go jsonrpc.StartRPCServer()
-	go websocket.StartServer()
-	if config.Parameters.HttpInfoPort > 0 {
-		go nodeinfo.StartServer(p2p.GetNetWork())
+	flag := false
+	select {
+	case <-exitCh:
+		if !flag {
+			return err
+		}
+	case <-time.After(time.Millisecond * 5):
+		flag = true
 	}
+	log.Infof("Rpc init success")
+	return nil
+}
 
-	go logCurrBlockHeight()
+func initLocalRpc(ctx *cli.Context) error {
+	if !ctx.GlobalBool(utils.RPCLocalEnableFlag.Name) {
+		return nil
+	}
+	var err error
+	exitCh := make(chan interface{}, 0)
+	go func() {
+		err = localrpc.StartLocalServer()
+		close(exitCh)
+	}()
 
-	//等待退出信号
-	waitToExit()
+	flag := false
+	select {
+	case <-exitCh:
+		if !flag {
+			return err
+		}
+	case <-time.After(time.Millisecond * 5):
+		flag = true
+	}
+
+	log.Infof("Local rpc init success")
+	return nil
+}
+
+func initRestful(ctx *cli.Context) error {
+	if !ctx.GlobalBool(utils.RestfulEnableFlag.Name) {
+		return nil
+	}
+	go restful.StartServer()
+
```

Comment:
given the function always return `nil`,  why the function signature return `error`?